### DISCUSSION TITLE
Dataset label

### DISCUSSION THREAD

**Discussion by DuanhaoranCC** [pyg-team/pytorch_geometric#4039] [General]
Hello, we are using the benchmark dataset provided by pyg（ https://pytorch_geometric.readthedocs.io/en/latest/modules/datasets.html# ), I want obtain the real label of the dataset, not 0 ~ n, but its real label semantics. What should I do if I want to conduct other analysis? No relevant information seems to be found in the document. thank you!

**Comment by rusty1s** [comment]
About which dataset are you talking about?

**Reply by DuanhaoranCC** [reply]
Basically all dataset, for example, I'm studying the mapping relationship between Cora's label and semantics (paper category). Thank you very much for your answer.

**Reply by rusty1s** [reply]
I think for this, you will need to look up the original source of the dataset. Some of these datasets are however only available in a pre-processed format where one no longer has access to the semantics of the labels/features, see [here](https://github.com/kimiyoung/planetoid#preprocessed-datasets).

**Reply by DuanhaoranCC** [reply]
OK, I see. Thank you.
Another problem is that I am rewriting the MessagePassing module to reproduce the complex aggregation method of custom(RNN etc.), but I find that MessagePassing only supports summation, averaging, pooling, etc., so I can only process one node by one, as shown in the following code, but the writing performance is a little low. Is there any problem with this writing. Do you have any suggestions? thank you.
```python
class GraphSage(MessagePassing):
    def __init__(self, in_channels, out_channels):
        super(GraphSage, self).__init__(aggr="add")
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.lin = nn.Linear(in_channels, out_channels)

    def forward(self, x, edge_index):
        # x has shape [N, in_channels], edge_index has shape [2, E]
        # Step 1: Add self-loops to the adjacency matrix.
        edge_index, _ = add_self_loops(edge_index, num_nodes=x.size(0))

        # Step 2: Linearly transform node feature matrix.
        x = self.lin(x)

        # Step 3-5: Start propagating messages.
        return self.propagate(edge_index, x=x)

    # Step 3
    def message(self, x_i, x_j, edge_index):
        # x_i neighbor node features, x_j source node features
        v = torch.FloatTensor([])
        for i in range(edge_index[0].max().item()+1):
            node = x_i[torch.eq(edge_index[0], i)]
            neighbor = node[0:-1]
            own = node[-1]
            # update...
            v = torch.cat([v, update], dim=0)
        v = v.view(-1, self.out_channels)
        return v

    # Step 5
    def update(self, aggr_out):
        return aggr_out

    # Step 4
    def aggregate(self, inputs, index, ptr, dim_size):
        # out = torch_scatter.scatter(inputs, index, dim=self.node_dim, dim_size=dim_size, reduce='sum')
        return inputs
```

**Reply by rusty1s** [reply]
Assuming that your edge indices are sorted column-wise (`torch_geometric.utils.sort_edge_index(..., sort_by_row=False)`), I would suggest you try to convert messages into shape `[num_nodes, num_max_neighbors, num_features]` to implement a custom aggregation function, e.g., via `torch_geometric.utils.to_dense_batch(...)`:
```python
def message(self, x_i, x_j, index):
    out = to_dense_batch(x_j, index)
    ...
```
